File: engine/src/Wsr2.Engine/Rng.cs
// Deterministic, seedable PRNG. Pure C# port of src/rng.ts (xoroshiro128**
// seeded from splitmix64). The same seed MUST produce the same uint64 stream
// as the TypeScript prototype — see RngParityTests for the parity contract.
//
// Required properties (per AGENTS.md §2):
//   * reproducible 30-year golden simulations,
//   * bit-exact snapshot/replay,
//   * server-authoritative multiplayer determinism.
//
// The cursor (state) is exposed as RngState so it can be snapshotted to JSON
// and restored later without losing fidelity.

using System.Numerics;

namespace Wsr2.Engine;

/// <summary>
/// Snapshot of a <see cref="Rng"/> cursor. <see cref="S0"/> and <see cref="S1"/>
/// are stored as decimal strings so the snapshot is JSON-safe and matches the
/// TypeScript prototype's wire format exactly.
/// </summary>
public sealed record RngState(string S0, string S1);

/// <summary>
/// xoroshiro128** PRNG, seeded via splitmix64. Produces the same stream as
/// the TS prototype's <c>Rng</c> for any given seed.
/// </summary>
public sealed class Rng
{
    private ulong _s0;
    private ulong _s1;

    public Rng(long seed) : this(unchecked((ulong)seed)) { }

    public Rng(ulong seed)
    {
        ulong s = seed;
        _s0 = SplitMix64Next(ref s);
        _s1 = SplitMix64Next(ref s);
        if (_s0 == 0 && _s1 == 0)
        {
            _s0 = 1;
        }
    }

    public Rng(RngState state)
    {
        ArgumentNullException.ThrowIfNull(state);
        _s0 = ulong.Parse(state.S0, System.Globalization.CultureInfo.InvariantCulture);
        _s1 = ulong.Parse(state.S1, System.Globalization.CultureInfo.InvariantCulture);
    }

    /// <summary>xoroshiro128** core. Returns a 64-bit unsigned integer.</summary>
    public ulong Next64()
    {
        ulong s0 = _s0;
        ulong s1 = _s1;
        ulong result = BitOperations.RotateLeft(s0 * 5UL, 7) * 9UL;
        s1 ^= s0;
        _s0 = BitOperations.RotateLeft(s0, 24) ^ s1 ^ (s1 << 16);
        _s1 = BitOperations.RotateLeft(s1, 37);
        return result;
    }

    /// <summary>Uniform double in [0, 1) with 53-bit precision.</summary>
    public double NextDouble()
    {
        // 2^53 fits exactly in a double; (Next64() >> 11) is at most 2^53 - 1.
        // The result is bit-identical to the TS prototype, which performs the
        // same arithmetic in IEEE 754 doubles.
        return (Next64() >> 11) / (double)(1UL << 53);
    }

    /// <summary>Uniform integer in [0, n). n must be a positive int.</summary>
    public int NextInt(int n)
    {
        if (n <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(n), n, "Rng.NextInt: n must be positive");
        }
        return (int)Math.Floor(NextDouble() * n);
    }

    /// <summary>Snapshot the current cursor for later restore.</summary>
    public RngState Snapshot() => new(
        _s0.ToString(System.Globalization.CultureInfo.InvariantCulture),
        _s1.ToString(System.Globalization.CultureInfo.InvariantCulture));

    private static ulong SplitMix64Next(ref ulong state)
    {
        state += 0x9E3779B97F4A7C15UL;
        ulong z = state;
        z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9UL;
        z = (z ^ (z >> 27)) * 0x94D049BB133111EBUL;
        return z ^ (z >> 31);
    }
}

File: src/rng.ts
/**
 * Deterministic, seedable PRNG. We use a small splitmix64-seeded xoroshiro128**
 * implementation in pure JS (BigInt) so the same seed produces the same stream
 * across Node versions and platforms. This is required for:
 *   - reproducible 30-year golden simulations,
 *   - bit-exact snapshot/replay,
 *   - server-authoritative multiplayer determinism.
 *
 * The cursor (state) is exposed so it can be snapshotted and restored.
 */

const MASK64 = (1n << 64n) - 1n;

function rotl(x: bigint, k: bigint): bigint {
  return ((x << k) | (x >> (64n - k))) & MASK64;
}

function splitmix64(seed: bigint): { next: () => bigint } {
  let s = seed & MASK64;
  return {
    next(): bigint {
      s = (s + 0x9e3779b97f4a7c15n) & MASK64;
      let z = s;
      z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & MASK64;
      z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & MASK64;
      return (z ^ (z >> 31n)) & MASK64;
    },
  };
}

export interface RngState {
  s0: string; // BigInt as string for JSON-safe snapshotting
  s1: string;
}

export class Rng {
  private s0: bigint;
  private s1: bigint;

  constructor(seed: number | bigint | RngState) {
    if (typeof seed === "object") {
      this.s0 = BigInt(seed.s0);
      this.s1 = BigInt(seed.s1);
      return;
    }
    const sm = splitmix64(typeof seed === "bigint" ? seed : BigInt(seed));
    this.s0 = sm.next();
    this.s1 = sm.next();
    if (this.s0 === 0n && this.s1 === 0n) this.s0 = 1n;
  }

  /** xoroshiro128** core. Returns a 64-bit BigInt. */
  private next64(): bigint {
    const s0 = this.s0;
    let s1 = this.s1;
    const result = (rotl((s0 * 5n) & MASK64, 7n) * 9n) & MASK64;
    s1 ^= s0;
    this.s0 = (rotl(s0, 24n) ^ s1 ^ ((s1 << 16n) & MASK64)) & MASK64;
    this.s1 = rotl(s1, 37n);
    return result;
  }

  /** Uniform float in [0, 1). 53-bit precision. */
  nextFloat(): number {
    const v = this.next64() >> 11n; // 53 bits
    return Number(v) / 2 ** 53;
  }

  /** Uniform integer in [0, n). n must be a positive safe integer. */
  nextInt(n: number): number {
    if (!Number.isInteger(n) || n <= 0) {
      throw new Error(`Rng.nextInt: n must be a positive integer, got ${n}`);
    }
    return Math.floor(this.nextFloat() * n);
  }

  /** Snapshot current cursor for later restore. */
  snapshot(): RngState {
    return { s0: this.s0.toString(), s1: this.s1.toString() };
  }
}
